backend_typescript/server/gameApi.ts

```diff
@@ -0,0 +1,326 @@
+    } catch(e) {
+        res.status(500).json({msg: 'Server error'});
+    }
+});
+
+
+// Image Upload & Retrieve
+const imageUpload = multer({
+    storage: multer.diskStorage(
+        {
+            destination: function (req, file, cb) {
+                cb(null, 'images/');
+            },
+            filename: function (req, file, cb) {
+                cb(
+                    null,
+                    new Date().valueOf() + 
+                    '_' +
+                    file.originalname
+                );
+            }
+        }
+    ), 
+});
+// Image Upload Routes
+gameRouter.post('/image',
+    imageUpload.single('image'),
+    async (req, res) => {
+        try {
+            const result = await db.insertImage(req.file.filename, req.file.path, req.file.mimetype, req.file.size);
+            res.status(200).json(result);
+        } catch (e) {
+            console.log(e);
+            res.status(500).json({ msg: 'Server error' });
+        }
+});
+// Image Get Routes
+gameRouter.get('/image/:filename', (req, res) => {
+    const { filename } = req.params;
+    const dirname = path.resolve();
+    const fullfilepath = path.join(dirname, 'images/' + filename);
+    return res.sendFile(fullfilepath);
+});
+
+
+// Create a new game
+gameRouter.post('/games', async (req, res: Response) => {
+    const {opponent, dimension}: GameBody = req.body;
+    const user = req.user as User;
+
+    if (dimension !== 10 && dimension !== 20) {
+        return res.status(400).json({msg: 'Dimension must be 10 or 20'});
+    }
+
+    if (opponent === user.username) {
+        return res.status(400).json({msg: 'Opponent must not be you'});
+    }
+
+    const gameManager = user.getGameManager();
+    const result = await gameManager.createGame(opponent, dimension);
+
+    if (result instanceof Failure) {
+        return res.status(result.status).json({msg: result.msg});
+    }
+    return res.status(201).json({msg: 'Game created'});
+});
+
+
+// Get all user games
+gameRouter.get('/games', async (req, res: Response) => {
+    const user = req.user as User;
+    const gameManager = user.getGameManager();
+    const results = await gameManager.getUserGames();
+    if (results instanceof Failure) {
+        return res.status(results.status).json({msg: results.msg});
+    }
+    const finalResults = results as Success<FinalResults>;
+    return res.status(200).json(finalResults.result);
+});
+
+
+// Update game- accept state
+gameRouter.put('/games/:game_id', async (req, res: Response) => {
+    const user = req.user as User;
+    const game = req.game as Game;
+    const result = await game.acceptGame(user.id);
+    if (result instanceof Failure) {
+        return res.status(result.status).json({msg: result.msg});
+    }
+    return res.status(200).json({msg: 'game accepted by opponent'});
+});
+
+
+// delete a game
+gameRouter.delete('/games/:game_id', async (req, res: Response) => {
+    const user = req.user as User;
```

Comment:
Maybe we can get rid of these annoying casts everywhere. I have an idea, but I'm not sure it will work. We can try it over Zoom.